src/hdmf/common/resources.py

```diff
@@ -304,22 +358,28 @@ def add_keys(self, **kwargs):
         Add key to be used for making references to external resources. This must be a DataFrame with the
         following columns:
             - *key_name*:              the key that will be used for referencing an external resource
-            - *resource_name*:         the name of the external resource
-            - *resource_entity_jd*:    the identifier for the entity at the external resource
-            - *resource_entity_uri*:   the URI for the entity at the external resource
+            - *resources_idx*:         the index for the resourcetable
+            - *entity_id*:    the index for the entity at the external resource
+            - *entity_uri*:   the URI for the entity at the external resource
 
-        It is possible to use the same *key_name* to refer to different resources so long as the *key_name* is not
-        used within the same object and field. This method does not support such functionality. See *add_key* and
+        It is possible to use the same name of the key to refer to different resources
+        so long as the name of the key is not used within the same object and field.
+        This method does not support such functionality. See *add_key* and
         *add_resource*.
         """
         res_df = popargs('res_df', kwargs)
         keys = res_df['key_name'].values
         ret = dict()
         for key in np.unique(keys):
             mask = keys == key
-            ret[key] = self.add_key(key)
-            for row in res_df[mask][[d['name'] for d in self.resources.__columns__[1:]]].to_dict('records'):
-                self.add_resource(ret[key], *row.values())
+            try:
+                self.get_key(key_name=key)
+            except Exception:
+                ret[key] = self.add_key(key)
+            else:
+                ret[key] = self.get_key(key_name=key)
```

Comment:
This seems inefficient and a catch-all `except Exception` is generally discouraged. I think this would also work:
```suggestion
            if key_name in self.keys:
                ret[key] = self.get_key(key_name=key)
            else:
                ret[key] = self.add_key(key)
```